compiler/qsc_rir/src/passes/reindex_qubits.rs

```diff
@@ -0,0 +1,177 @@
+        // Assume qubits only appear in void call instructions.
+        match instr {
+            Instruction::Call(call_id, args, _)
+                if program.get_callable(*call_id).call_type == CallableType::Reset =>
+            {
+                // Generate any new qubit ids and skip adding the instruction.
+                for arg in args {
+                    if let Value::Literal(Literal::Qubit(qubit_id)) = arg {
+                        qubit_map.insert(*qubit_id, next_qubit_id);
+                        next_qubit_id += 1;
+                    }
+                }
+            }
+            Instruction::Call(call_id, args, None) => {
+                // Map the qubit args, if any, and copy over the instruction.
+                let new_args = args
+                    .iter()
+                    .map(|arg| match arg {
+                        Value::Literal(Literal::Qubit(qubit_id)) => {
+                            if let Some(mapped_id) = qubit_map.get(qubit_id) {
+                                highest_used_id = highest_used_id.max(*mapped_id);
+                                Value::Literal(Literal::Qubit(*mapped_id))
+                            } else {
+                                *arg
+                            }
+                        }
+                        _ => *arg,
+                    })
+                    .collect::<Vec<_>>();
+
+                // If the call was to mresetz, replace with mz.
+                let call_id = if Some(*call_id) == mresetz_id {
+                    used_mz = true;
+                    mz_id
+                } else {
+                    *call_id
+                };
+
+                new_block.push(Instruction::Call(call_id, new_args, None));
+
+                if program.get_callable(call_id).call_type == CallableType::Measurement {
+                    // Generate any new qubit ids after a measurement.
+                    for arg in args {
+                        if let Value::Literal(Literal::Qubit(qubit_id)) = arg {
+                            qubit_map.insert(*qubit_id, next_qubit_id);
+                            next_qubit_id += 1;
+                        }
+                    }
+                }
+            }
+            _ => {
+                // Copy over the instruction.
+                new_block.push(instr.clone());
+            }
+        }
+    }
+
+    program.num_qubits = highest_used_id + 1;
+    program.blocks.clear();
+    program.blocks.insert(block_id, Block(new_block));
+
+    // All reset function calls should be removed, so remove them from the callables.
+    let mut callables_to_remove = Vec::new();
+    for (callable_id, callable) in program.callables.iter() {
+        if callable.call_type == CallableType::Reset || Some(callable_id) == mresetz_id {
+            callables_to_remove.push(callable_id);
+        }
+    }
+    for callable_id in callables_to_remove {
+        program.callables.remove(callable_id);
+    }
```

Comment:
```rust
program.callables.retain(|(callable_id, callable)| {
                callable.call_type != CallableType::Reset && Some(callable_id) != mresetz_id
            });
```

Reply:
oh, nice, thanks!

Reply:
Darn, I'll have to add an impl for that to `IndexMap` since it's not defined yet. Probably worth it though.